File: kakin_ex/Chapter08/Chapter08/Chapter8/BlockTable.swift
//
//  BlockTable.swift
//  Chapter8
//
//  Copyright © 2016年 shoeisya. All rights reserved.
//

import Foundation

class BlockTable {
    static let blockColNum = 5
    static let blockRowNum = 3
    
    var blocks:[Block] = []
    var availableBlockNum: Int
    
    init(pattern: [Bool], width: Float, height: Float) {
        availableBlockNum = 0
        
        // ブロック1つ分の横幅、高さを計算する
        let blockWidth = width / (Float(BlockTable.blockColNum) + 2)
        let blockHeight = height / 15
        
        // ブロックを生成する
        for i in 0..<BlockTable.blockRowNum {
            for j in 0..<BlockTable.blockColNum {
                let status = pattern[i * BlockTable.blockColNum + j]
                
                let block = Block.init(x: blockWidth + Float(j) * blockWidth,
                                       y: blockHeight + Float(i) * blockHeight,
                                       width: blockWidth,
                                       height: blockHeight,
                                       status: status)
                blocks.append(block)
                
                if status {
                    // ブロックテーブル上に表示されるブロック数を数える
                    availableBlockNum += 1
                }
            }
        }
    }
    
    func setBall(_ ball:Ball) {
        blocks.forEach{ (block) in
            if block.status == false {
                return
            }
            
            // ボールがブロックの底辺に当たった場合
            if (block.x < ball.x - ball.radius && ball.x + ball.radius < block.x + block.width) &&
                (block.y + block.height / 2.0 < ball.y - ball.radius && ball.y - ball.radius < block.y + block.height) {
                // ボールのY軸進行方向を反転する
                ball.changeDirectionY()
                
                // ブロックを非表示にする
                block.status = false
                
                // ブロックの有効数を1つ減らす
                availableBlockNum -= 1
            }
                // ボールがブロックの上辺に当たった場合
            else if (block.x < ball.x - ball.radius && ball.x + ball.radius < block.x + block.width) &&
                (block.y < ball.y + ball.radius && ball.y + ball.radius < block.y + block.height / 2.0) {
                // ボールのY軸進行方向を反転する
                ball.changeDirectionX()
                
                // ブロックを非表示にする
                block.status = false
                
                // ブロックの有効数を1つ減らす
                availableBlockNum -= 1
            }
                // ボールがブロックの右辺に当たった場合
            else if (block.x + block.width - 10 < ball.x - ball.radius && ball.x - ball.radius < block.x + block.width) &&
                (block.y < ball.y - ball.radius && ball.y + ball.radius < block.y + block.height) {
                // ボールのX軸進行方向を反転する
                ball.changeDirectionX()
                
                // ブロックを非表示にする
                block.status = false
                
                // ブロックの有効数を1つ減らす
                availableBlockNum -= 1
            }
                // ボールがブロックの左辺に当たった場合
            else if (block.x < ball.x + ball.radius && ball.x + ball.radius < block.x + 10) &&
                (block.y < ball.y - ball.radius && ball.y + ball.radius < block.y + block.height) {
                // ボールのX軸進行方向を反転する
                ball.changeDirectionX()
                
                // ブロックを非表示にする
                block.status = false
                
                // ブロックの有効数を1つ減らす
                availableBlockNum -= 1
            }
        }
    }
}

File: kakin_ex/Chapter04/Chapter04/app/src/main/java/jp/co/se/androidkakin/Chapter04/BlockTable.java
package jp.co.se.androidkakin.Chapter04;

import java.util.ArrayList;

/*
 * ブロックテーブルを表現するクラス
 */
public class BlockTable {
	// ブロックテーブルを表現するArrayList
	private ArrayList<Block> blocks;
	
	// ブロックテーブル上に表示されるブロック数
	private int availableBlockNum;
	
	// コンストラクタ
	public BlockTable(boolean pattern[], int width, int height) {
		blocks = new ArrayList<Block>();
		availableBlockNum = 0;
		
		int block_width = width / (Utils.BLOCK_COL_NUM + 2); // ブロック1つ分の横幅を計算する
		int block_height = height / 15; // ブロック1つ分の高さを計算する
		
		for (int i=0; i<Utils.BLOCK_ROW_NUM; i++) {
			for (int j=0; j<Utils.BLOCK_COL_NUM; j++) {
				blocks.add(new Block(block_width+(j*block_width), block_height+(i*block_height), block_width, block_height, 0, 255, 0, pattern[i*Utils.BLOCK_COL_NUM+j]));
				if (pattern[i*Utils.BLOCK_COL_NUM+j]) {
					availableBlockNum++; // ブロックテーブル上に表示されるブロック数を数える
				}
			}
		}
	}
	
	// ブロックを取得する
	public Block getBlock(int number) {
		return blocks.get(number);
	}
	
	// ブロック数を取得する
	public int getSize() {
		return blocks.size();
	}

	// ボールを設定する
	public void setBall(Ball ball) {
		for (int j=0; j<blocks.size(); j++) {
			Block block = blocks.get(j);
			
			if (block.getStatus() == false) {
				continue;
			}
			
			// ボールがブロックの底辺に当たった場合
			if ((block.getX() < ball.getX() && ball.getX() < block.getDx()) && ((ball.getY() < block.getDy()+10) && (ball.getY() > block.getDy()))) {
				// ボールのY軸進行方向が上向きだった場合は下向きに設定する
				if (ball.getDirectionY() == Utils.BALL_DIRECTION_TOP) {
					ball.setDirectionY(Utils.BALL_DIRECTION_BOTTOM);
				// ボールのY軸進行方向が下向きだった場合は上向きに設定する
				} else {
					ball.setDirectionY(Utils.BALL_DIRECTION_TOP);
				}

				// ブロックを非表示にする
				block.setStatus(false);

				// ブロックの有効数を1つ減らす
				availableBlockNum--;
			}
			// ボールがブロックの上辺に当たった場合
			else if ((block.getX() < ball.getX() && ball.getX() < block.getDx()) && ((ball.getY() > block.getY()-10) && (ball.getY() < block.getY()))) {
				// ボールのY軸進行方向が上向きだった場合は下向きに設定する
				if (ball.getDirectionY() == Utils.BALL_DIRECTION_TOP) {
					ball.setDirectionY(Utils.BALL_DIRECTION_BOTTOM);
				// ボールのY軸進行方向が下向きだった場合は上向きに設定する
				} else {
					ball.setDirectionY(Utils.BALL_DIRECTION_TOP);
				}

				// ブロックを非表示にする
				block.setStatus(false);

				// ブロックの有効数を1つ減らす
				availableBlockNum--;
			}
			// ボールがブロックの右辺に当たった場合
			else if ((block.getDx() < ball.getX() && ball.getX() < block.getDx()+10) && ((ball.getY() > block.getY()) && (ball.getY() < block.getDy()))) {
				// ボールのX軸進行方向が右向きだった場合は左向きに設定する
				if (ball.getDirectionX() == Utils.BALL_DIRECTION_RIGHT) {
					ball.setDirectionX(Utils.BALL_DIRECTION_LEFT);
				// ボールのX軸進行方向が左向きだった場合は右向きに設定する
				} else {
					ball.setDirectionX(Utils.BALL_DIRECTION_RIGHT);
				}

				// ブロックを非表示にする
				block.setStatus(false);

				// ブロックの有効数を1つ減らす
				availableBlockNum--;
			}
			// ボールがブロックの左辺に当たった場合
			else if ((block.getX()-10 < ball.getX() && ball.getX() < block.getX()) && ((ball.getY() > block.getY()) && (ball.getY() < block.getDy()))) {
				// ボールのX軸進行方向が右向きだった場合は左向きに設定する
				if (ball.getDirectionX() == Utils.BALL_DIRECTION_RIGHT) {
					ball.setDirectionX(Utils.BALL_DIRECTION_LEFT);
				// ボールのX軸進行方向が左向きだった場合は右向きに設定する
				} else {
					ball.setDirectionX(Utils.BALL_DIRECTION_RIGHT);
				}

				// ブロックを非表示にする
				block.setStatus(false);

				// ブロックの有効数を1つ減らす
				availableBlockNum--;
			}
		}
	}
	
	// 残っているブロック数を取得する
	public int getAvailableBlockNum() {
		return availableBlockNum;
	}
}
